### PR TITLE
[Feature:TAGrading] Clear version conflicts with save

### PR DESCRIPTION
**Pull request by dorian451**
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [ ] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
When a component has a version conflict, hitting the save button without changing the marks doesn't do anything.

### What is the new behavior?
The save button will now update a component's graded version on save. It will tell the user beforehand.

#### Before:
2024-07-15T10:32:50,492398213-04:00

#### After:
2024-07-15T10:36:24,159880711-04:00

### CODE REVIEW

File: site/public/js/ta-grading-rubric.js [hunk 1]
```diff
@@ -3414,6 +3419,7 @@ function injectInstructorEditComponentHeader(component, showMarkList) {
  */
 function injectGradingComponent(component, graded_component, editable, showMarkList) {
     const student_grader = $('#student-grader').attr('is-student-grader');
+    console.log('rendering:', graded_component);
```
Comment (williamjallen) [inline site/public/js/ta-grading-rubric.js:3422]: Is this left over from your local testing?

File: site/public/templates/grading/SavingTools.twig [hunk 2]
```diff
@@ -11,8 +11,8 @@ Required Parameters:
         <span class="btn btn-default save-tools-cancel key_to_click" tabindex="0" onclick="{{ cancel_on_click }} ;event.stopPropagation()">
             Discard
         </span>
-        <span class="btn btn-primary save-tools-save" data-testid="save-tools-save">
-            Save
+        <span class="btn btn-primary save-tools-save">
+            Save {% if component_version_conflict %} (and resolve version conflict) {% endif %}
```
Comment (williamjallen) [inline site/public/templates/grading/SavingTools.twig:15]:
```suggestion
            Save {% if component_version_conflict %} and resolve version conflict {% endif %}
```

**Review by williamjallen — COMMENTED** [review]
It would be good for @IDzyre to take a look at the new Cypress test too.

**Comment by williamjallen** [comment]
Can you please add screenshots of the before/after functionality here?

File: site/public/js/ta-grading-rubric.js (hunk 1 above)
Comment (dorian451) [reply site/public/js/ta-grading-rubric.js:3422]: yeah oops

**Comment by dorian451** [comment]
Just uploaded

File: site/cypress/e2e/Cypress-TAGrading/clearing_version_conflicts.spec.js [hunk 3]
```diff
@@ -38,4 +38,47 @@ describe('Test cases for checking the clear version conflicts button in the TA g
...
         cy.get('[data-testid="version-warning"]').should('exist');
     });
+
+    it('Clicking "save" on a component with a version conflict should resolve the conflict for that component', () => {
+        cy.login('instructor');
+        cy.visit(['sample', 'gradeable', 'grading_homework', 'grading', 'grade?who_id=K8jI3q4qpdCc1jw&sort=id&direction=ASC&gradeable_version=1']);
+        cy.get('#grading_rubric_btn').click();
+
+        // wait until page is fully loaded
+        cy.get('#component-list').children().should('have.length.least', 1);
```
Comment (IDzyre) [inline site/cypress/e2e/Cypress-TAGrading/clearing_version_conflicts.spec.js:48]: Could you use testids for the elements tested in this PR?

**Review by IDzyre — CHANGES_REQUESTED** [review]

File: site/cypress/e2e/Cypress-TAGrading/clearing_version_conflicts.spec.js (hunk 3 above)
Comment (dorian451) [reply site/cypress/e2e/Cypress-TAGrading/clearing_version_conflicts.spec.js:48]: My bad I thought I changed them already

**Review by zacharymnp — APPROVED** [review]
works as intended

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 4]
```diff
@@ -2365,6 +2365,78 @@ public function saveGradedComponent(TaGradedGradeable $ta_graded_gradeable, Grad
         }
     }
 
+    /**
+     * Route for saving the marks the submitter received for a component
+     */
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/graded_gradeable/change_grade_version", methods: ["POST"])]
+    public function ajaxUpdateGradedVersionForStudent(string $gradeable_id): ?Jsonresponse {
+        $anon_id = $_POST['anon_id'] ?? null;
+        $graded_version = intval($_POST['graded_version'] ?? null);
```
Comment (williamjallen) [inline site/app/controllers/grading/ElectronicGraderController.php:2374]: If you're going to cast it to an int anyway, why not just use an int for the default value?

```suggestion
        $graded_version = intval($_POST['graded_version'] ?? 0);
```

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 5]
```diff
@@ -2365,6 +2365,78 @@ public function saveGradedComponent(TaGradedGradeable $ta_graded_gradeable, Grad
         }
     }
 
+    /**
+     * Route for saving the marks the submitter received for a component
+     */
+    #[Route("/courses/{_semester}/{_course}/gradeable/{gradeable_id}/grading/graded_gradeable/change_grade_version", methods: ["POST"])]
+    public function ajaxUpdateGradedVersionForStudent(string $gradeable_id): ?Jsonresponse {
```
Comment (williamjallen) [inline site/app/controllers/grading/ElectronicGraderController.php:2372]: A route should never return null.  A proper response should always be returned.  Also, it looks like you have a typo in the return type name.

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 6]
```diff
@@ -2365,6 +2365,78 @@ public function saveGradedComponent(TaGradedGradeable $ta_graded_gradeable, Grad
...
+
+        // Get the gradeable
+        $gradeable = $this->tryGetGradeable($gradeable_id);
+        if ($gradeable === false) {
+            return null;
+        }
+        // Get user id from the anon id
+        $submitter_id = $this->tryGetSubmitterIdFromAnonId($anon_id, $gradeable_id);
+        if ($submitter_id === false) {
+            return null;
```
Comment (williamjallen) [inline site/app/controllers/grading/ElectronicGraderController.php:2396]: See comment about never returning null in a user-facing route.

File: site/app/templates/grading/electronic/RubricPanel.twig [hunk 7]
```diff
@@ -27,9 +27,18 @@
...
-                            <input id="verify-all" type="button" class="btn btn-default key_to_click" tabindex="0" value="Verify All" onclick="onVerifyAll(this)"/>
+                            <input id="verify-all" type="button" class="btn btn-default key_to_click mx-1" tabindex="0" value="Verify All" onclick="onVerifyAll(this)"/>
                         </div>
                     {% endif %}
+
+                    {# clear version conflicts button #}
+                    {% if show_clear_conflicts and not version_conflict %}
+                        <div class="col-no-gutters">
+                            <label for="change-graded-version" tabIndex="0" class="screen-reader">Clear Version Conflicts</label>
+                            <input id="change-graded-version" data-testid="change-graded-version" type="button" class="btn btn-default key_to_click mx-1" tabindex="0" value="Clear Version Conflicts" onclick="updateAllComponentVersions()"/>
```
Comment (williamjallen) [inline site/app/templates/grading/electronic/RubricPanel.twig:38]: This line is pretty long.  Would it be better to put each attribute on a separate line?

**Review by williamjallen — CHANGES_REQUESTED** [review]

File: site/app/controllers/grading/ElectronicGraderController.php (hunk 5 above)
Comment (dorian451) [reply site/app/controllers/grading/ElectronicGraderController.php:2372]: Other routes in the same file also return `null` in the same situation, and also seem to be user-facing routes (like `ajaxSetPeerFeedback`), so I assumed it was a special case or something.

File: site/app/controllers/grading/ElectronicGraderController.php (hunk 6 above)
Comment (dorian451) [reply site/app/controllers/grading/ElectronicGraderController.php:2396]: https://github.com/Submitty/Submitty/pull/10679#discussion_r1681591609

File: site/app/templates/grading/electronic/RubricPanel.twig (hunk 7 above)
Comment (dorian451) [reply site/app/templates/grading/electronic/RubricPanel.twig]: Addressed in 87ed1b98d9eb3e92a3bc44dd336c61bb7de3357c

File: site/app/controllers/grading/ElectronicGraderController.php (hunk 6 above)
Comment (dorian451) [reply site/app/controllers/grading/ElectronicGraderController.php]: Resolved with https://github.com/Submitty/Submitty/pull/10679/commits/db4c3dc26fbb4f4c089c58fc7d586238593afdc8

File: site/app/controllers/grading/ElectronicGraderController.php (hunk 5 above)
Comment (dorian451) [reply site/app/controllers/grading/ElectronicGraderController.php]: Resolved with https://github.com/Submitty/Submitty/pull/10679/commits/db4c3dc26fbb4f4c089c58fc7d586238593afdc8

**Review by IDzyre — APPROVED** [review]
Tests look good to me, haven't tested the functionality more than just what the tests do.

**Review by oliiso — APPROVED** [review]
Functionality works well and UI is clear.

**Review by williamjallen — APPROVED** [review]
The code is much cleaner now and looks reasonable to me.

**Pull request merged**